src/plugins/chatMarkers/ChatMarkersPlugin.ts

```diff
@@ -0,0 +1,410 @@
+   private sendReceived(lastReceivedMsgId: string) {
+      this.pluginAPI.send($msg({
+         to: this.data.get('lastReceivedFrom')
+      }).c(RECEIVED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastReceivedMsgId
+      }));
+   }
+
+   // send "displayed" message according to XEP-0333
+   private sendDisplayed(lastDisplayedMsgId: string) {
+      this.pluginAPI.send($msg({
+         to: this.data.get('lastDisplayedFrom')
+      }).c(DISPLAYED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastDisplayedMsgId
+      }));
+   }
+
+   // send "acknowledged" message according to XEP-0333
+   private sendAcknowledged(lastAcknowledgedMsgId: string) {
+      this.pluginAPI.send($msg({
+         to: this.data.get('lastAcknowledgedFrom')
+      }).c(ACKNOWLEDGED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastAcknowledgedMsgId
+      }));
+   }
+
+   private setMessagesChatMarkersAcknowledged(lastAcknowledgedMsgUid: string) {
+      let contact = this.pluginAPI.getContact(new JID(this.data.get('lastAcknowledgedFrom')));
+      if (contact !== undefined) {
+         let transcript = contact.getChatWindow().getTranscript();
+         let msg = transcript.getMessage(lastAcknowledgedMsgUid);
+         while (!!msg && !msg.isChatMarkersAcknowledged()) {
+            if (msg.getDirection() === DIRECTION.IN) {
+               msg.chatMarkersAcknowledged();
+            }
+            try {
+               msg = transcript.getMessage(msg.getNextId());
+            } catch (error) {
+               break;
+            }
+         }
+      }
+   }
+
+   private preSendMessageStanzaProcessor = (msg: Message, stanza: Strophe.Builder): Promise<any> => {
+      return this.supportsChatMarkers(msg.getPeer()).then(() => {
+         if (msg.getType() === Message.MSGTYPE.CHAT) {
+            this.addMarkable(stanza, msg);
+         }
+         return Promise.resolve([msg, stanza]);
+      });
+   }
+
+   private preSendMessageProcessor = (contact: Contact, msg: Message): Promise<any> => {
+      return this.supportsChatMarkers(msg.getPeer()).then(() => {
+         if (msg.getType() === Message.MSGTYPE.CHAT) {
+            this.registerOutHooks(msg, contact.getChatWindow().getChatWindowMessage(msg));
+         }
+         return Promise.resolve([contact, msg]);
+      });
+   }
+
+   private afterReceiveMessageProcessor = (contact: Contact, msg: Message, stanza: Element): Promise<any> => {
+      return this.supportsChatMarkers(msg.getPeer()).then(() => {
+         if (msg.getType() === Message.MSGTYPE.CHAT) {
+            this.addInIndicators(msg, contact.getChatWindow().getChatWindowMessage(msg));
+            this.registerInHooks(msg, contact.getChatWindow().getChatWindowMessage(msg));
+         }
+         return Promise.resolve([contact, msg, stanza]);
+      });
+   }
+
+   // send "received" on "markable" message according to XEP-0333
+   private onMarkableMessage = (stanza: string) => {
```

Comment:
Section 9 says "It is possible for a sender to leak its presence when updating Chat Markers; therefore, a sender SHOULD NOT send Chat Markers to recipients who are not otherwise authorized to view its presence.". How is this ensured?

Reply:
>  How is this ensured?

I still see no solution for this.